source/common/http/conn_manager_impl.cc

```diff
@@ -1074,6 +1077,26 @@ void ConnectionManagerImpl::ActiveStreamDecoderFilter::
   parent_.connection_manager_.stats_.named_.downstream_flow_control_resumed_reading_total_.inc();
 }
 
+void ConnectionManagerImpl::ActiveStreamDecoderFilter::addDownstreamWatermarkCallbacks(
+    DownstreamWatermarkCallbacks& watermark_callbacks) {
+  // This is called exactly once per stream, by the router filter.
+  // If there's ever a need for another filter to subscribe to watermark callbacks this can be
+  // turned into a vector.
+  ASSERT(parent_.watermark_callbacks_ == nullptr);
+  parent_.watermark_callbacks_ = &watermark_callbacks;
+  for (uint32_t i = 0; i < parent_.high_watermark_count_; ++i) {
+    watermark_callbacks.onAboveWriteBufferHighWatermark();
+  }
+}
+void ConnectionManagerImpl::ActiveStreamDecoderFilter::removeDownstreamWatermarkCallbacks(
+    DownstreamWatermarkCallbacks& watermark_callbacks) {
+  ASSERT(parent_.watermark_callbacks_ == &watermark_callbacks);
+  for (uint32_t i = 0; i < parent_.high_watermark_count_; ++i) {
```

Comment:
Yeah, I'm concerned about this too.

So let's say we get rid of it.  For the H2 case, it's entirely possible that an individual stream gets deleted without undoing its readDisables().  Since readDisable results in not consuming window and we have safeguards to consume when the stream is closed we're OK.  

I'm unsure of the logical place to do the same for HTTP though.  We'd need to make sure that if upstream sent a ton of data, blocking the downstream socket, the upstream connection was done and detached and reused, it wouldn't stay write blocked.  Can you suggest somewhere to make sure we don't end up assigning a non-reading Network::Connection to another request?